src/protocols/rdp/rdp.c

```diff
@@ -525,12 +525,24 @@ static int guac_rdp_handle_connection(guac_client* client) {
     /* Set default pointer */
     guac_common_cursor_set_pointer(rdp_client->display->cursor);
 
+    /* 
+     * Downgrade the lock to allow for concurrent read access
+     * Access to read locks need to be made available for other processes to use 
+     * while we await credentials from the user.
+     */
+    pthread_rwlock_unlock(&(rdp_client->lock));
```

Comment:
I like this approach, but I think there's at least one assignment that should probably still be guarded by the write lock - namely:

https://github.com/apache/guacamole-server/blob/main/src/protocols/rdp/rdp.c#L112

The lock itself is a bit unclear on exactly what it's guarding, and it _might_ be ok to have this particular assignment unguarded since it's only set once, so the only stale value that any other thread/process could read would be `NULL`.

Reply:
Maybe it would make sense to guard just that one assignment using the write lock? Possibly using a `guac_rwlock`?

Thoughts @mike-jumper?

Reply:
On one hand I like the idea of adding a separate lock for around setting the audio handler, but on the other hand I'm not keen on nesting more locks

Reply:
Sorry, I was unclear here. What I meant was using `guac_rwlock` to convert the entire `rdp_client->lock` into a reentrant rwlock, so you could keep the current read lock around the `freerdp_connect()` call, and then additionally grab the write lock, but only when the `rdp_client->audio_input` (or any assignment) is done.

Reply:
Oh I just looked through the guac_rwlock api and that seems much cleaner. I can try and implement that

Reply:
Cool, looks good. I realize now that this doesn't _really_ require reentrancy (acquiring the write lock while holding the read lock, as you're doing now is equivalent to just releasing the read lock and acquiring the write lock).

But it certainly won't hurt to use the reentrant locks either.